lib/bitvec.cpp

```diff
@@ -145,3 +146,20 @@ bool bitvec::is_contiguous() const {
         return false;
     return max().index() - min().index() + 1 == popcount();
 }
+
+/**
+ * The purpose of this function is to barrel_shift a bitvec around a pivot.  This function
+ * will crash if the pivot is smaller than the size of the bitvector
+ */
+bitvec bitvec::barrel_shift_left(size_t shift, size_t pivot) const {
```

Comment:
It seems this is a rotate, not a barrel shift.  Might be more cleanly/clearly expressed as such, with 3 args (rotate sz bits at offset idx by shift).

Reply:
suggest:

    // In place rotate (left) of sz bits starting at idx by shift
    bitvec &rotate(size_t idx, size_t sz, size_t shift) {
        bitvec bits = getslice(idx, sz);
        bits <<= idx + shift % sz;
        bits |= bits >> (sz - shift % sz);
        clrrange(idx, sz);
        bit.clrrange(0, idx);
        bit.clrrange(idx + sz, ~0);
        *this |= bits;
        return *this;
    }

could be done not in place as well.